### ISSUE TITLE
Renaming files on desktop will change their position

### ISSUE THREAD

**Issue by Atalanttore** [lxde/lxqt#1030]
Hi,

the position of a file on desktop will be changed automatically after renaming it.

That's annoying (at least to me), but not a major problem.

Regards,
Ettore

**Comment by tsujan** [comment]
Right click on the desktop and open the Sorting sub-menu; you'll find the reason. This is the *correct* behavior not an issue.

**Comment by paulolieuthier** [comment]
@tsujan you are right, but ideally, for the desktop, the icon should be placed in the same spot as before being renamed. What do you think?

**Comment by tsujan** [comment]
> ideally, for the desktop, the icon should be placed in the same spot as before

I don't agree at all. Suppose you've renamed a lot of items. If that was the case, how could you find them?! Sorting is for finding items as fast as possible.

If you want an item to be in a fixed position as an exception, then right click → Stick to current position. If you rename it and still want it to be in a fixed position, you could repeat the same procedure. After all, how many times may one want to rename items?

**Comment by pmattern** [comment]
IMO this issue is due automatic grouping of items on the desktop *always* being active. This is something that doesn't make really sense, you may want to have a look at the notes I dropped in https://github.com/lxde/pcmanfm-qt/pull/341#issuecomment-213545295.

**Comment by tsujan** [comment]
@pmattern 
OK! That's about a separate *feature request* and I think its implementation is possible, although perhaps not easy. https://github.com/lxde/pcmanfm-qt/pull/341 is about the lack of reading and saving of sorting for desktop, i.e. about a *bug*.

**Comment by tsujan** [comment]
@pmattern
In addition, you said, in https://github.com/lxde/pcmanfm-qt/pull/341#issuecomment-213545295,

> When "QupZilla" gets dragged and dropped from the main menu to some arbitrary location on the desktop it will end in the upper left corner of the desktop (most of the time, when exactly?!). But when Firefox gets dragged & dropped the same way it will end at the upper left corner as well replacing QupZilla due to the sorting by ascending name.

That's *not* due to the sorting by ascending name but by ascending mod. time. Your QupZilla's desktop file is probably modified *before* other items on your desktop, and your Firefox's desktop file before QupZilla's. So, the placement is completely logical.

Being aware of the fact that I'm only a newcomer in LXQT and that most members have a much longer history than me, I beg everyone for these:

(1) Don't presuppose that everything unexpected to you is a bug (= don't be pessimistic ;) );
(2) Don't mistake feature requests for bugs; 
(3) Don't report or discuss other bugs in PRs; and
(4) Give priority to fixing current bugs rather than adding new features.

IMHO, development will be faster and easier if these conditions are fulfilled. Please think of them as a friendly suggestion!

**Comment by Atalanttore** [comment]
@tsujan 
The explanatory phrase “It's not a bug, it's a feature” is not helpful for annoyed users of the current behaviour.

Regards,
Ettore

**Comment by tsujan** [comment]
> The explanatory phrase “It's not a bug, it's a feature” is not helpful for annoyed users of the current behaviour.

It wasn't for affecting anyone's emotions or feelings but to bring order to the process of developing.

**Comment by tsujan** [comment]
@Atalanttore
And where did I say "It's not a bug, it's a feature"?!! I said that it's about a feature *request* not a bug.

**Comment by tsujan** [comment]
@jleclanche Thanks for marking this as a discussion!

This discussion started when @Atalanttore said that renaming desktop items changed their position. Well, it does *not* do so here because my desktop items are ordered according to the modification time. Therefore, I told @Atalanttore to take a look at the Sorting sub-menu (and see that the items are sorted by name for him -- I didn't said that explicitly because it was obvious). Apparently, @Atalanttore got confused but didn't say why.

Then, @paulolieuthier asked my opinion about fixing the position on renaming and I disagreed because if the sorting is based on names, why shouldn't the position change after renaming.

Then @pmattern mentioned two different things: a PR for fixing a bug, and a valid feature request. He also thought that the latter was related to what was said before.

Then, it was *my* turn to get confused by all this ;) Therefore, I tried to bring some order to the discussion -- the same thing I'm trying to do now.

**Comment by jleclanche** [comment]
To be clear, you're correct this isn't a bug. But we probably don't want this as default behaviour. Being able to manually order icons is a pretty basic and expected feature.

**Comment by Vladimir-csp** [comment]
The whole issue can be described in a more simple way: contrary to a normal file manager window, desktop icon management predisposes to treating sorting as one-time action, unless specific 'constant sorting' option is preferred by user.

**Comment by tsujan** [comment]
@jleclanche, @Vladimir-csp, @pmattern 
Thanks! This is getting clearer a little, but still not enough. Let me be more specific:

As I see it, *sorting is inevitable* because when an item is before or after another one, the question naturally arises why that's so. Therefore, the real question is not about the removal of sorting but about adding another kind of it. As far as I can understand from your comments, you may mean sorting based on the time of creation. Am I right?

**Comment by jleclanche** [comment]
By default, we should not do any sorting if the items are already ordered. If the order hasn't been touched, we can sort by whatever flavour works best (by name sounds fine).

**Comment by Vladimir-csp** [comment]
> As far as I can understand from your comments, you may mean sorting based on the time of creation. Am I right?

No. All the same kinds of sorting, but on-demand, rather than continuous. Not "keep em' sorted", but "sort and forget".

**Comment by tsujan** [comment]
> By default, we should not do any sorting if the items are already ordered.

There's always an order unless an item is stuck to some position. I could rewrite the above sentence as, "we should not do any sorting if the items are stuck", and that's what already happens.

Practically speaking, suppose I've installed PCManFM for the first time and want to run it with its desktop module. There are already several items in my desktop folder. The module is started and it should have been given a way to put the items in a grid: by name, size, type, modification time, ..., creation time or even with a random seed (which is a very impractical way). Which way do you mean? You cannot say, "no way!"

**Comment by tsujan** [comment]
I mean, don't think of desktop as a real object that was already there with some items in it, before starting PCManFM!

**Comment by jleclanche** [comment]
@tsujan That is the default state, which is "untouched", which is autosorted. As soon as the order is in place, it's sorted.

**Comment by tsujan** [comment]
> That is the default state, which is "untouched",

What could be the default state in a grid? Vague concepts don't help here. We should write lines of code.

**Comment by jleclanche** [comment]
@tsujan It doesn't matter. Like I said, by name should be fine.

**Comment by tsujan** [comment]
@jleclanche Please read my question at https://github.com/lxde/lxqt/issues/1030#issuecomment-213731569 and give a specific answer to it! Is this what you mean: the first time, PCManFM should order items based on the name and the other times based on the creation time? That would be a mess! Which time is the first time? Why should we care when a user runs PCManFM for the first time in his/her whole life? What's the relevance?

**Comment by jleclanche** [comment]
I'm not sure how this is confusing, it's been default behaviour on Windows/Mac OS going back decades.

**Comment by Vladimir-csp** [comment]
Regarding the special case of launching desktop module, If there is a saved icons placement, do nothing. If there is no saved icons placement, perform sorting by something (name seems appropriate).

**Comment by tsujan** [comment]
> I'm not sure how this is confusing, it's been default behaviour on Windows/Mac OS going back decades.

OK! As far as I remember, Windows places items based on the creation time, but when I mentioned it, @Vladimir-csp said no and no one said anything.

> If there is a saved icons placement, do nothing. If there is no saved icons placement, perform sorting by something (name seems appropriate). 

That's exactly the current behavior of PCManFM. "saved icons placement" is about the items stuck to their positions by the user.

**Comment by Vladimir-csp** [comment]
> That's exactly the current behavior of PCManFM. "saved icons placement" is about the items stuck to their positions by the user.

No, currently PCManFM maintains continuous sorting on desktop, just like in normal FM window (with the exception of pinned icons).

**Comment by LStranger** [comment]
>No, currently PCManFM maintains continuous sorting on desktop, just like in normal FM window (with the exception of pinned icons).

There is no such thing as "continuous sorting", the desktop manager just
discovers there is new file (irrelevant was it added or renamed one) then
it decides where to put it, and that decision is based on sorting method
used, which usually is by time. Although when fixed position item gets
renamed it should maintain its fixed position, and if that never happens
then that's a bug. Otherwise it's just expected behavior in dependence on
settings.

**Comment by tsujan** [comment]
@Vladimir-csp Maybe this is what you mean: The first time PCManFM is run, it sorts items based on the default scheme (name, size, etc.) and *then*, considers all of them pinned. When another item is added, it's also added to the end of the pinned items list *automatically*. When one of the existing (pinned) items is renamed, it remains pinned and its place in the list isn't changed but when it's deleted, it's also deleted from the list. All this can be called as sorting based on creation time on desktop.

@LStranger  Not changing the position of stuck items on renaming seems a good idea to me.

**Comment by Vladimir-csp** [comment]
> The first time PCManFM is run, it sorts items based on the default scheme (name, size, etc.) and then, considers all of them pinned. When another item is added, it's also added to the end of the pinned items list automatically.

Basically, yes. But that saved state should be different than pinned so pinned items would retain their positions when user decides to perform sorting manually via menu.

**Comment by tsujan** [comment]
> ...But that saved state should be different than pinned...

So, two kinds of pinning, one based on the position (which already exists) and the other based on what? You said it shouldn't be based on the creation time on desktop. I'm baffled!

A concept may be taken for granted at first because it may seem "natural" but unless its algorithmic structure isn't found, it couldn't be translated into lines of code.

Whatever Linux desktop I've had, sorting based on modification time was the best for me (I hated the Windows desktop mess 13 years ago). My proposals were rejected and, apparently, there isn't a clear consensus yet. So, I'll wait for one and follow it.

**Comment by Vladimir-csp** [comment]
OK, here is an algorithm:
**Cold start, nothing was saved before**: perform sorting (based on default sort mode, i.e. name), write positions list. Each entry of the list should contain: name, pin state, position.
**New icon appears**: put it on first unoccupied position, add entry to the list.
**File renamed**: change entry's name.
**Icon removed**: remove entry.
**User moves icon**: change entry's position.
**User pins icon**: set entry's pin state true.
**User unpins icon**: set entry's pin state false.
**User renames pinned file**: change entry's name.
**User moves pinned icon**: change entry's position.
**User orders sort** (by selecting sort mode in context menu): sort all icons, except pinned, according to selected mode, update positions in their entries.
**Restart** (list already exists): remove entries of missing files from the list, restore positions according to list, add entries for new files placing them on first available positions.

**Comment by tsujan** [comment]
@Vladimir-csp Clear enough for a start. What about this: I try to implement my understanding of what you said in a private branch and you test the result. If *the result* is accepted by you and others, it'll just remain to polish the code. If not, everyone testing it could *specifically* tell why. No installation will be needed -- just a straightforward compilation and running as `./pcmanfm-qt --desktop --profile=default`, for example, after stopping desktop from Session Settings. What do you think?

**Comment by Vladimir-csp** [comment]
I agree.

**Comment by tsujan** [comment]
Great :) It may just take some time (due to priority of other fixes/PRs, difficulty of the job, my free time ...) but I'll do it and tell everyone here.

**Comment by pmattern** [comment]
I still think omitting any automatic sorting of items on the desktop should be possible, at least when it comes to dragging & dropping items from panel's main menu to the desktop which is a pretty common proceeding.

Background / use case: The vast majority of desktop users I know want to arrange items on their desktop in a way that's independent from the features the underlying files have in their capacity as FHS object, namely file name, mtime and the likes. Users may e. g. want to arrange all internet applications in a row at the upper left, sorted by frequency of personal usage, multimedia applications in a row below and so on.
At least on a new user profile with an empty desktop it's pretty difficult to arrange items dragged and dropped from the panel menu that way as long as automatic sorting is active.

**Comment by tsujan** [comment]
> ... at least when it comes to dragging & dropping items from panel's main menu to the desktop...

Yes, that's really easy. The problem is finding the best way to save and restore the order. I'll work on it.

**Comment by gCurse** [comment]
It seems pcmanfm (gtk) handles the desktop nearly the way that is requested here.
Maybe that code could be a starting point?

**Comment by tsujan** [comment]
I succeeded in implementing what's discussed here by some "acrobatics" and without touching libfm-qt (no testable branch yet) but I like to discuss it more before continuing the job in that direction:

When you get the properties of a file in Dolphin, it shows *the creation time* too, while PCManFM doesn't. If you paste any file into the desktop, for example, its creation time will be the time of pasting. So, it occurred to me (again) that if we could add that to libfm-qt, then we'll need to do nothing other than adding an item to the Sorting menu with this text: "By Creation Time".

@Vladimir-csp why do you think this isn't what you want? I ask you because the others didn't say "no" when I proposed it. And, please, avoid using over-generalized concepts in your reply!

**Comment by Vladimir-csp** [comment]
AFAIK, creation time isn't being tracked by main Linux filesystems.
Anyway, sorting by creation time would just be another sorting mode added to the others. It does not have any direct connection to the algorithm described in https://github.com/lxde/lxqt/issues/1030#issuecomment-213750684

**Comment by tsujan** [comment]
@Vladimir-csp Please tell me *the effects* of which steps in that algorithm are violated by sorting according to the creation time!

**Comment by Vladimir-csp** [comment]
Sorting by creation time does not violate any part of that algorithm, it is just one mode among others.

**Comment by tsujan** [comment]
If, for the user, *the effect* is the same (which isn't but it might be with some modifications), then it'll make no difference how we call it: another sorting mode or whatever. On the other hand, the code will be much cleaner.

P.S. These are just speculations. I'm almost sure that the direction I took with the above-mentioned "acrobatics" will give acceptable results while I'm not sure about creation time yet. However, the latter seems more promising and I'll try it too.

**Comment by Vladimir-csp** [comment]
I fail to understand why sorting by creation time has some special significance here, but won't do any harm anyway.

**Comment by tsujan** [comment]
> ...why sorting by creation time has some special significance here

With the current direction, the ordering should be saved to a file and restored from it on restarting PCManFM because it doesn't fit into any category and is specific to the user's activities. It's not a big deal but it'll be better and cleaner if there's no need to that -- both on the coding level and logically. Anyhow, sorting by creation time is worth a try.

**Comment by tsujan** [comment]
To be as clear as possible, I should add this:

Like windowed folders, desktop is made based on `QSortFilterProxyModel`, which means that *sorting is inevitable*. I don't say that removal of sorting is impossible but it requires fundamental changes in libfm-qt -- and that would not only be a tedious job but also senseless for such a minor feature.

Therefore, only two possibilities remain: sorting based on creation time and sorting based on a saved list. However, for the above reasons and AFAIK, there's no way to distinguish between renaming and removing+creating. That means, even with the list approach, if a file is renamed, it'll go to the end. (It also means that there's no way to keep the "Stick to current position" property after renaming.) Therefore, the creation-time and list approaches result in one and the same effect.

This brings us back to the first step. The discussion started with a mild complaint about why renamed desktop items changed their positions and my answer was they would NOT if the user chose sorting by  modification time (which was the default until a recent commit).

Frankly speaking, I'm somehow demotivated by the amount of misunderstandings here and from now on, I won't participate in vague discussions. Although I finished my work based on the list approach, I'll try to add sorting based on the creation time later because it -- and only it -- is clear enough and makes sense, IMHO.

**Comment by Vladimir-csp** [comment]
The algorithm I described earlier is basically sorting by list, with occasional on-demand re-sorting on user's request.
If sorting by list is already possible, then it becomes the matter of updating the list in the right moment of time. Is there a way to grab old and new file name from rename dialog and update the list before next sorting iteration?

**Comment by tsujan** [comment]
> ... basically sorting by list, with occasional on-demand re-sorting on user's request.... If sorting by list is already possible, then it becomes the matter of...

I did all that and it worked but (1) AFAIK, there's no way to know which file is renamed and how because renaming can be done in various ways and there isn't a general signal for it in libfm-qt; (2) with regard to (1), this is exactly sorting by creation time but done in a rediculous way (by needlessly saving a list).

**Comment by Vladimir-csp** [comment]
It is safe to assume that most renaming events on desktop would be renaming via PCManFM desktop component itself, so grabbing names in its dialog would be sufficient. Other cases are understandably out of scope.

**Comment by tsujan** [comment]
@Vladimir-csp
*All* renaming is done by PCManFM -> libfm-qt -> libfm. Renaming can be done by shortcut, right-clicking and Properties dialog. Only the first one is in control of PCManFM.

**Comment by tsujan** [comment]
Ah, only if libfm-qt used `QFileSystemModel`, we would have the nice signal `QFileSystemModel::fileRenamed()`, which would give all the info we needed!

I haven't found a workaround yet.

**Comment by tsujan** [comment]
OK, I finished my long investigation by concluding that this is impossible with the current structure. The LXQt desktop is *designed* to be sorted from the start. And even if we had a "renaming signal", the renamed file would first move to the end and then to the position of the original file, i.e. an ugly jump would occur.

If you want (unpinned) items not to change their positions on renaming, sort them according to the modification time.

What *is* possible with the current structure is adding a sorting-by-creation-time, which wouldn't keep the positions of renamed items but would result in dropped items being positioned after the current ones.

**Comment by tsujan** [comment]
This is moved to https://github.com/lxde/pcmanfm-qt/issues/500. Please comment on that page!
